### DISCUSSION TITLE
Proposal: Allow T! or T-? to fullfill where T  : notnull

### DISCUSSION THREAD

**Discussion by TheConstructor** [dotnet/csharplang#6968] [Language Ideas]
Currently, when you write a generic class/method/... that uses a type like ConcurrentDictionary that constraints one of it generic parameters to be `notnull`, you will not get the warnings in the most helpful places:

```cs
public class DictionarySink<TKey, TValue>
{
	// Compiler warning: 'CS8714: Nullability of type argument doesn't match 'notnull' constraint.'
	private readonly ConcurrentDictionary<TKey, TValue> _dictionary = new();

	public void Accept(TKey key, TValue value)
	{
		if (key != null)
		{
			// We checked -> all fine
			_dictionary[key] = value;
		}
		/*
		// Just for completeness, writing the below doesn't currently raise a warning, even though now clearly the where TKey : notnull of ConcurrentDictionary is violated
		else
		{
			// should emit a warning like 'CS8604: Possible null reference argument for a parameter.' - but does not!
			_dictionary[key] = value;
		}
		*/
	}
}
```

I would like to propose `T!` or `T-?` (similar to TypeScript) to remove a possible nullability of the generic parameter. The resulting code would then look like that, with only a single warning:

```cs
public class DictionarySink<TKey, TValue>
{
	// No warning, as TKey! is specified, lifting a possible ? passed to define TKey by a user of this class
	private readonly ConcurrentDictionary<TKey!, TValue> _dictionary = new();

	public void Accept(TKey key, TValue value)
	{
		if (key != null)
		{
			_dictionary[key] = value;
		}
		/*
		// Just for completeness, using a key, that is not checked to be notnull, should raise a warning e.g.
		else
		{
			// Compiler warning: 'CS8604: Possible null reference argument for a parameter.'
			_dictionary[key] = value;
		}
		*/
	}
}
```

Why does this help? The declaration of the `ConcurrentDictionary` doesn't actually cause problems, but only the passing in of `null`s. While I'd argue, that `CS8604` should already be raised by now, as the class type is constraint as `notnull`, simply suppressing the `CS8714`-warnings doesn't improve anything. Using `T!` or `T-?` I can state my intent of handling `null`-values without passing them to the derived class.

This could also offer a new way to write my favorite nullability-adjusting-filter:
```cs
public static IEnumerable<T!> WhereNotNull<T>(this IEnumerable<T> source)
{
	return source.Where(x => x is not null)!;
}
```

**Comment by HaloFour** [comment]
C# already has syntax for declaring this constraint:

```cs
public class DictionarySink<TKey, TValue> where TKey : notnull
{
     // ...
}
```

You could also define `Accept` as generic and apply the constraint, so that it warns appropriately there:

```cs
public void Accept(TKey2 key, TValue value) where TKey2 : notnull, TKey {
    // ...
}
```

**Reply by TheConstructor** [reply]
No, that's totally the opposite of what I want to do. I want to be able to use `DictionarySink<Something?, SomethingElse?>` knowing, that if there are no warnings in `DictionarySink`-code all possible `null`s have been accounted for, before a call to `ConcurrentDictionary` is made. Think of e.g. an optional cache.

**Reply by TheConstructor** [reply]
Tried to clarify the example code

**Comment by TahirAhmadov** [comment]
OK after you clarified your OP code example, it makes more sense, and it probably falls under the "runtime restriction of generic constraints" umbrella. Basically, people have been also asking for syntax like `if (T is IEnumerable) { foreach(var x in t) { ... } }`. This would probably be added alongside those improvements.

**Comment by Joe4evr** [comment]
Duplicate of #6308

**Reply by TheConstructor** [reply]
Definitely related, but not sure if it's an exact duplicate. Will likely take a closer look later

**Reply by Joe4evr** [reply]
You're asking for the ability to have one method with an unconstrained (or *under*constrained) type parameter call another method with a (more) constrained type parameter. It's an exact duplicate.

**Reply by TheConstructor** [reply]
I want to express a constraint change for a field/property. Then I want to use it after checking the constraint on the value. The first sentence seemed missing on first glance
